### PR TITLE
Fix rounded corners on speaker's video

### LINKED ISSUE #8156: Video of other person in 1on1 call is not rounded
### How to use GitHub

* Please use the 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to show that you are affected by the same issue.
* Please don't comment if you have no relevant information to add. It's just extra noise for everyone subscribed to this issue.
* Subscribe to receive notifications on status change and new comments.

---

## Steps to reproduce
1. Start a call
2. Look at the video of the other person

### Expected behaviour
Nice border-radius, just like for your own video.

### Actual behaviour
Sharp edges, ow!
image

## Talk app

**Talk app version:** Latest on cloud.nc.com

### PR DESCRIPTION
**Pull request by Antreesy**
Signed-off-by: Maksim Sukharev <antreesy.web@gmail.com>

Fix #8156 

### 🖼️ Screenshots

🏚️ Before | 🏡 After
---|---
One-to-One call | 
Screenshot from 2023-02-03 10-52-49 | Screenshot from 2023-02-03 10-48-16
Group call | 
Screenshot from 2023-02-03 10-51-57 |  Screenshot from 2023-02-03 10-49-37

- [X] Fulscreen view is rounded as well
- [X] Screencast isn't affected
- [X] Video in portrait mode is not cutted

### 🚧 TODO

- [ ] Code-review
- [ ] Visual check

### 🏁 Checklist

- [X] 📘 API documentation in `docs/` has been updated or is not required
- [X] 📗 User documentation in https://github.com/nextcloud/documentation/tree/master/user_manual/talk has been updated or is not required
- [X] 🔖 Capability is added or not needed

### CODE REVIEW

**Review by ShGKme — COMMENTED** [review]
It also affects the layout of a portrait. I'd prefer not to have the video cut.

🏚️ Before | 🏡 After
---|---
image | image

**Comment by nickvergessen** [comment]
> I'd prefer not to have the video cut.

Yeah that is actually a requirement from product management

**Comment by Antreesy** [comment]
Ok, looking for a better solution

**Review by ShGKme — COMMENTED** [review]
Adding ResizeObserver to round a border seems too overhead to me...

I haven't tried it on the flexbox, don't know if this is possible. But it is definitely possible with the gridbox layout.

**Comment by Antreesy** [comment]
> But it is definitely possible with the gridbox layout.

I'm afraid that this is going to be the same trick for the `grid` as for the `flex` (see below), and these are issues:
- both require additional wrapper for `<video>` element;
- working only for portait mode, where container's `width` param is usually less then `height`;
- padding is calculated from parent width, so the total size is probably going to be larger then we need.

```
.videoAdditionalWrapper {
  width: 100%;
  height: 0;
  padding-top: 56.25%; // reverted aspect-ratio, can be obtained from eventListener
  position: relative;
}
.video {
  width: 100%;
  height: 100%;
  position: absolute;
  top: 0;
  left: 0;
}
```
So, if there is a more simple solution, i'll be glad to hear it.

For now, we can restrict ResizeObserver and EventListener to be called only for speaker's frame (which is only one at a time) to reduce memory consumption. Unfortunately, i couldn't find a better way to get `aspect-ratio` of video stream and orientation of container.

**Comment by ShGKme** [comment]
We could get `aspect-ratio` from 

```js
model.attributes.stream.getVideoTracks()[0].getSettings().aspectRatio
```

But yep. It doesn't work with gridbox either... 

I cannot make it at the same time stretched + centralized + kept aspect ratio... (´。＿。｀)

Not very beautiful solution: use the fact that header's and paddings' heights are known and fixed.

**Comment by ShGKme** [comment]
There is a working solution using [aspect-ratio](https://developer.mozilla.org/en-US/docs/Web/CSS/aspect-ratio). It has been supported in modern browsers since ~Jun..Sep 2021

File: src/components/CallView/shared/VideoVue.vue [hunk 1]
```diff
@@ -475,6 +490,14 @@ export default {
...
 		this._setStream(this.model.attributes.stream)
+
+		if (this.isBig) {
+			this.$refs.video.addEventListener('playing', this.getVideoAspectRatio)
+		}
+	},
+
+	beforeDestroy() {
+		this.$refs.video.removeEventListener('playing', this.getVideoAspectRatio)
 	},
```
Comment (ShGKme) [inline src/components/CallView/shared/VideoVue.vue:501]: In Vue component's elements events are usually set with `v-on` directive in the template. Then all component's DOM events would be defined in the template and we wouldn't need to remove listeners manually.

```html
<video @playing="updateVideoAspectRatio"></video>
```

By the way, do we need to handle this event to update the aspect ratio? I'm not sure, but it seems to me that anytime the video is changed the stream would be also changed. Then the `model.attributes.stream` watcher with `_setStream` handler will be triggered. So we may calculate the aspect ratio immediately on the stream receiving from `model.attributes.stream.getVideoTracks()[0].getSettings().aspectRatio` while attaching the media stream before mounting to the DOM.

Comment (Antreesy) [reply src/components/CallView/shared/VideoVue.vue:501]: Unfortunately, this method doesn't return aspectRatio neither after setting stream, nor after `nextTick` call (see below):

image


these params appear in the settings only when video is on, which is - the best i could find - right after `playing` event
```
{
    "aspectRatio": 1.3333333333333333,
    "frameRate": 30,
    "height": 360,
    "width": 480
}
```

**Review by ShGKme — APPROVED** [review]
Looks good to me, tested in Chromium.
But I'm not sure about Safari, as it is kind of a new feature in CSS...

https://caniuse.com/mdn-css_properties_aspect-ratio

I want to test it on Mac later today.

**Review by ShGKme — CHANGES_REQUESTED** [review]
It doesn't work on Safari, the video is always stretch like `width: 100%`

(┬┬﹏┬┬)

image

image

**Comment by Ivansss** [comment]
Tested and works in Safari 👍

**Review by ShGKme — APPROVED** [review]
It's sad to realize that in 2023 there is still no solution for this problem in CSS without calculating the size manually in JS ಥ_ಥ

Thanks for all this work and attempts.

**Pull request merged**